### PR TITLE
/lib/ukfallocbuddy: Add tests for ukfallocbuddy lib

### PR DESCRIPTION
**Pull request by flpostolache**
Add some basic tests for the ukfallocbuddy library.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): `x86_64`
 - Platform(s): `kvm`
 - Application(s): `N/A`

### Additional configuration
 - `CONFIG_LIBUKFALLOCBUDDY_TEST=y`

### Description of changes

Add some basic tests for the `ukfallocbuddy` library.

### CODE REVIEW

File: lib/ukfallocbuddy/Config.uk [hunk 1]
```diff
@@ -26,4 +26,18 @@ config LIBUKFALLOCBUDDY_STATS
...
 	default n
 
+config LIBUKFALLOCBUDDY_TEST
+	bool "Enable tests"
+	default n
+	select LIBUKTEST
+
+if LIBUKFALLOCBUDDY_TEST
+
+config LIBUKFALLOCBUDDY_PUBLIC_FUNCTIONS
```
Comment (StefanJum) [inline lib/ukfallocbuddy/Config.uk]: What exactly is the purpose of the `LIBUKFALLOCBUDDY_PUBLIC_FUNCTIONS`?
From what I can see it does not seem to do anything, did I miss something?

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+struct uk_falloc *generate(void *fa_meta, void *start_zone, unsigned long len, __paddr_t *new_start);
+
+
+UK_TESTCASE(ukfallocbuddy, test_init)
+{
+	// test if the init function works
+	struct uk_falloc *fa;
+
+	fa = malloc(100000);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: I think we should use the `uk_fallocbuddy_size()` function when allocating the required memory for the
`struct uk_falloc` pointer, as specified [here](https://github.com/unikraft/unikraft/blob/staging/lib/ukfallocbuddy/include/uk/fallocbuddy.h#L48).

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	UK_TEST_EXPECT_SNUM_EQ(uk_fallocbuddy_init(fa), 0);
+	free(fa);
+}
+
+UK_TESTCASE(ukfallocbuddy, test_size)
+{
+	UK_TEST_EXPECT_SNUM_EQ(uk_fallocbuddy_size(), 368);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Is this always going to be 368?

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+{
+	// test if the addmem functionality works
+	void *start_zone = malloc(10000);
+	__sz fa_meta_size;
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	__sz len = 10000;
+	unsigned long pages = len >> PAGE_SHIFT;
+	void *fa_meta = malloc(10000);
+	struct uk_falloc *fa = malloc(100000);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: See the comment above about allocating the required space using `uk_fallocbuddy_size()`.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+}
+uk_testsuite_register(ukfallocbuddy, NULL);
+
+struct uk_falloc *generate(void *fa_meta, void *start_zone, unsigned long len, __paddr_t *new_start)
+{
+	__sz fa_meta_size;
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	unsigned long pages = len >> PAGE_SHIFT;
+	struct uk_falloc *fa = malloc(100000);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: See the comment above about allocating the required space using `uk_fallocbuddy_size()`.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	// test allocation with too many pages
+	void *start_zone = malloc(50000);
+	void *fa_meta = malloc(10000);
+	__paddr_t start;
+	__sz len = 50000;
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+	int falloc_result;
+
+	falloc_result = fa->falloc(fa, (__paddr_t *)start, 600, 0);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_result, -14);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
	UK_TEST_EXPECT_SNUM_EQ(falloc_result, -EFAULT);
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	__paddr_t start;
+	__paddr_t end;
+	__sz len = 50000;
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+	int falloc_from_range_result;
+
+	end = start + 8192;
+	fa->falloc(fa, (__paddr_t *)start, 2, 0);
+	falloc_from_range_result = fa->falloc_from_range(fa, (__paddr_t *)start, 2, 0, start, end);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, -12);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, -ENOMEM);
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 3 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: It's true. The result is not constant. I will remove it.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 2 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Done.

File: lib/ukfallocbuddy/Config.uk (hunk 1 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/Config.uk]: Initially, I proposed to @marcrittinghaus  there should be 2 sources that contain tests for `ukfallocbuddy` library. One should contain tests with the functions that are in the `fallocbuddy.h` file and one contains tests with all static functions that are present in the `ukfallocbuddy/fallocbuddy.c`. After looking at it better I saw no need add one more. I've erased it now.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+UK_TESTCASE(ukfallocbuddy, test_add_memory)
+{
+	// test if the addmem functionality works
+	void *start_zone = malloc(10000);
+	__sz fa_meta_size;
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	__sz len = 10000;
+	unsigned long pages = len >> PAGE_SHIFT;
+	void *fa_meta = malloc(10000);
```
Comment (StefanJum) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Since we get the size of fallocbuddy metadata at line 35, in `fa_meta_size`, using `uk_fallocbuddy_metadata_size`,
should we use that when allocating memory for `fa_meta`?

Same for the other space allocations for the metadata below.

**Review by StefanJum — COMMENTED** [review]
I left a comment regarding allocating space for the allocator metadata.
All looks good besides that.

**Review by StefanJum — DISMISSED** [review]
All looks good now!
I think you can remove the `draft` tag so the tests can run.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 9]
```diff
@@ -0,0 +1,189 @@
+#include <uk/test.h>
+#include <uk/fallocbuddy.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c:19]:
```suggestion
```

Please remove any unnecessary extra blank lines. The same applies to the double blank lines between the `UK_TESTCASE`s below. and the extra newline added to the config file

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+static struct uk_falloc *generate(void *fa_meta, void *start_zone, unsigned long len,
+									__paddr_t *new_start, __sz fa_meta_size, unsigned long *pages)
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
static struct uk_falloc *generate(void *fa_meta, void *start_zone,
				  unsigned long len, __paddr_t *new_start,
				  __sz fa_meta_size, unsigned long *pages)
```
Configure your editor to 8 spaces per tab. Make sure that you properly align parameter lists and do not exceed 80 characters per column (please fix everywhere).

Also the function name `generate` is not really fitting, I think. Maybe something like `new_falloc` would be better.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	free(fa_meta);
+	free(fa);
+}
+
+UK_TESTCASE(ukfallocbuddy, test_check_variables)
+{
+	__paddr_t start;
+	__sz len = 10000;
+	unsigned long pages = len >> PAGE_SHIFT;
+	__sz fa_meta_size =  uk_fallocbuddy_metadata_size(pages);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
	__sz fa_meta_size = uk_fallocbuddy_metadata_size(pages);
```
There is an extra space here

**Comment by gabrielmocanu** [comment]
@marcrittinghaus  Any updates on this?

**Comment by marcrittinghaus** [comment]
> @marcrittinghaus Any updates on this?

There should have been some by now. Sorry. I will check this out again. Thanks for reminding 🙏🏻

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+
+
+static struct uk_falloc *generate(void *fa_meta, void *start_zone, unsigned long len,
+									__paddr_t *new_start, __sz fa_meta_size, unsigned long *pages)
+{
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	struct uk_falloc *fa = malloc(uk_fallocbuddy_size());
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Add a `UK_ASSERT` in cases you imply any particular value (in this case a successful allocation). The same applies to all the other functions here that return a value.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+	dm_off = start;
+	fa->addmem(fa, fa_meta, start, *pages, dm_off);
+	*new_start = start;
+	return fa;
+}
+
+
+UK_TESTCASE(ukfallocbuddy, test_init)
+{
+	// test if the init function works
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: We do not use `//` comments in Unikraft. Use `/* */` instead. You may also consider using a comment in front of the test case and there also describe the reasoning behind the test. For example, for the allocation tests it would be good to describe what corner cases you want to catch.

```C
/**
 * Tests if the init functions works by ...
 */
UK_TESTCASE(...
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+	int addmem_result;
+
+	uk_fallocbuddy_init(fa);
+
+	start = PAGE_ALIGN_UP(start + fa_meta_size);
+	pages = (len - fa_meta_size) >> PAGE_SHIFT;
+	dm_off = start;
+	addmem_result = fa->addmem(fa, fa_meta, start, pages, dm_off);
+	UK_TEST_EXPECT_SNUM_EQ(addmem_result, 0);
+	free(start_zone);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c:135]: You may add a blank line here to more optically separate the finalization from the test code

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+UK_TESTCASE(ukfallocbuddy, test_check_variables)
+{
+	__paddr_t start;
+	__sz len = 10000;
+	unsigned long pages = len >> PAGE_SHIFT;
+	__sz fa_meta_size =  uk_fallocbuddy_metadata_size(pages);
+	void *start_zone = malloc(len);
+	void *fa_meta = malloc(fa_meta_size);
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start, fa_meta_size, &pages);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: This code is used a couple of times in various forms. Consider creating a function for it to reduce duplication and to bring the actual test into the focus.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+	int falloc_from_range_result;
+
+	min = start + 4096;
+	max = start + 20000;
+	fa->falloc(fa, (__paddr_t *)start, 1, FALLOC_FLAG_ALIGNED);
+	falloc_from_range_result = fa->falloc_from_range(fa, (__paddr_t *)start, 2, 0, min, max);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: This is obviously longer than 80 cols.

**Review by marcrittinghaus — CHANGES_REQUESTED** [review]
Hi @maniatro111,

thanks again for submitting the test cases and sorry for taking so long to review! I have added a couple of comments mostly related to the formatting and structure of the code.

Concerning the test cases themselves, I think you already did a great job. Obviously, for a thorough testing, we would need to test more rigorously. For a software component as complex as an allocator it also makes sense to test internal behavior and check internal data structures for consistency. For this we could implement more tests in a header file and include this at the end of the allocator's C file. This way we would have access to static symbols and private declarations in the tests. But this is not needed for this PR.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 9 above)
Comment (marcrittinghaus) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c:19]: Empty line still here 😏

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+	unsigned long pages;
+	__sz fa_meta_size;
+	void *start_zone;
+	void *fa_meta;
+};
+
+
+static struct uk_falloc *init_new_fb_alloc(void *fa_meta, void *start_zone,
+	unsigned long len, __paddr_t *new_start, __sz fa_meta_size,
+	unsigned long *pages)
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Parameters should be aligned. Since this will not allow you to put more than one parameter per line it is fine to move the return value to the previous line.
```suggestion
static struct uk_falloc *
init_new_fb_alloc(void *fa_meta, void *start_zone, unsigned long len, 
		  __paddr_t *new_start, __sz fa_meta_size, unsigned long *pages)
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+	struct uk_falloc *fa;
+	__paddr_t start;
+	__sz len;
+	unsigned long pages;
+	__sz fa_meta_size;
+	void *start_zone;
+	void *fa_meta;
+};
+
+
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
```
Remove extra empty line

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+	UK_ASSERT(fa = malloc(uk_fallocbuddy_size()));
+
+	UK_ASSERT(!uk_fallocbuddy_init(fa));
+	start = PAGE_ALIGN_UP(start + fa_meta_size);
+	*pages = (len - fa_meta_size) >> PAGE_SHIFT;
+	dm_off = start;
+	UK_ASSERT(!fa->addmem(fa, fa_meta, start, *pages, dm_off));
+	*new_start = start;
+	return fa;
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: You cannot use `UK_ASSERT()` to perform any action, because the expression is not evaluated when asserts are disabled. I suggest you add the following helper macro:
```C
#define fab_bug_on(cond)						\
	do {								\
		if (unlikely(cond))					\
			UK_CRASH("'%s' during test execution.\n",	\
				 STRINGIFY(cond));			\
	} while (0)
```
And change the code to this:
```suggestion
	__paddr_t start = (__paddr_t)start_zone;
	struct uk_falloc *fa;
	int rc;

	fa = malloc(uk_fallocbuddy_size());
	fab_bug_on(fa == NULL);

	rc = uk_fallocbuddy_init(fa);
	fab_bug_on(rc != 0);

	start = PAGE_ALIGN_UP(start + fa_meta_size);
	*pages = (len - fa_meta_size) >> PAGE_SHIFT;

	rc = fa->addmem(fa, fa_meta, start, *pages, start);
	fab_bug_on(rc != 0);

	*new_start = start;
	return fa;
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+						test_attributes->start_zone,
+						test_attributes->len,
+						&(test_attributes->start),
+						test_attributes->fa_meta_size,
+						&(test_attributes->pages));
+	UK_ASSERT(test_attributes->fa);
+
+	return test_attributes;
+
+}
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: 1. Also here you do not want to rely on `UK_ASSERT()` for things that can go wrong at runtime. 
2. Separate the definition from the initialization for the attributes to make it fit in one line. Also change the name to make it more compact.
3. Add empty lines to group things semantically. For example, we here group all `fa_meta` stuff and make that apparent via the separation.
4. Remove the extra empty line after the return
5. Remove the extra parenthesis after the `&` operator. 
6. Change the name of the function to something more expressive.

```suggestion
static struct test_ukfallocbuddy *new_allocator(__sz len)
{
	struct test_ukfallocbuddy *attr;

	attr = (struct test_ukfallocbuddy *)malloc(sizeof(*attr));
	fab_bug_on(attr == NULL);

	attr->len = len;
	attr->pages = len >> PAGE_SHIFT;
	attr->start_zone = malloc(len);
	fab_bug_on(attr->start_zone == NULL);

	attr->fa_meta_size = uk_fallocbuddy_metadata_size(attr->pages);
	attr->fa_meta = malloc(attr->fa_meta_size);
	fab_bug_on(attr->fa_meta == NULL);

	attr->fa = init_new_fb_alloc(attr->fa_meta, attr->start_zone,
				     attr->len, &attr->start,
				     attr->fa_meta_size, &attr->pages);
	fab_bug_on(attr->fa == NULL);

	return attr;
}
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+static void free_memory(struct test_ukfallocbuddy *test_attr)
+{
+	free(test_attr->start_zone);
+	free(test_attr->fa_meta);
+	free(test_attr->fa);
+	free(test_attr);
+}
+
+/**
+ * Test if the init function works
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: According to the new coding convention we are putting together with the community, we would write here:
```suggestion
 * Tests if the init function works
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+
+	start = PAGE_ALIGN_UP(start + fa_meta_size);
+	pages = (len - fa_meta_size) >> PAGE_SHIFT;
+	dm_off = start;
+	addmem_result = fa->addmem(fa, fa_meta, start, pages, dm_off);
+	UK_TEST_EXPECT_SNUM_EQ(addmem_result, 0);
+
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: 1. You mostly reuse here code that you have split in `initialize_parameters()` and `init_new_fb_alloc()`. I suggest to write it here in the same style as before so you recognize it is the same code. This makes is easier to maintain.
2. I increased the memory size that you add. You were adding less than 2.5 pages. Although it would be valid to have a test that explores the lower and upper limits.
3. You could argue that - as a best practice - you should release memory in the order that there is no internal references (in this case the exact reverse). But I kept it as-is since it is all released anyways. But you keep this in mind in other cases.

```suggestion
	__sz len;
	unsigned long pages;
	void *start_zone;
	__sz fa_meta_size;
	void *fa_meta;
	struct uk_falloc *fa;
	__paddr_t start;
	int rc;

	len = 10 * 1024 * 1024; /* 10 MiB */
	pages = len >> PAGE_SHIFT;
	start_zone = malloc(len);
	fab_bug_on(start_zone == NULL);

	fa_meta_size = uk_fallocbuddy_metadata_size(pages);
	fa_meta = malloc(fa_meta_size);
	fab_bug_on(fa_meta == NULL);

	fa = malloc(uk_fallocbuddy_size());
	fab_bug_on(fa == NULL);

	rc = uk_fallocbuddy_init(fa);
	fab_bug_on(rc != 0);

	start = PAGE_ALIGN_UP((__paddr_t)start_zone + fa_meta_size);
	pages = (len - fa_meta_size) >> PAGE_SHIFT;

	rc = fa->addmem(fa, fa_meta, start, pages, start);
	UK_TEST_EXPECT_ZERO(rc);

	free(start_zone);
	free(fa_meta);
	free(fa);
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+/**
+ * Test a simple allocation
+ */
+UK_TESTCASE(ukfallocbuddy, test_simple_falloc)
+{
+	struct test_ukfallocbuddy *test = initialize_parameters(10000);
+	int falloc_result;
+
+	falloc_result = test->fa->falloc(test->fa, (__paddr_t *)test->start,
+						1, 0);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Broken indentation

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 24]
```diff
@@ -0,0 +1,217 @@
...
+UK_TESTCASE(ukfallocbuddy, test_alloc_from_range)
+{
+	struct test_ukfallocbuddy *test = initialize_parameters(50000);
+	__paddr_t min, max;
+	int falloc_from_range_result;
+
+	min = test->start + 4096;
+	max = test->start + 20000;
+	test->fa->falloc(test->fa, (__paddr_t *)test->start, 1,
+				FALLOC_FLAG_ALIGNED);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Broken indentation

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+/**
+ * Test allocation with too many pages
+ */
+UK_TESTCASE(ukfallocbuddy, test_alloc_large)
+{
+	struct test_ukfallocbuddy *test = initialize_parameters(50000);
+	int falloc_result;
+
+	falloc_result = test->fa->falloc(test->fa, (__paddr_t *)test->start,
+						600, 0);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Broken indentation

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 26]
```diff
@@ -0,0 +1,217 @@
...
+ * Test allocation of two frames between an already allocated one
+ */
+UK_TESTCASE(ukfallocbuddy, test_alloc_over_2)
+{
+	struct test_ukfallocbuddy *test = initialize_parameters(50000);
+	int falloc_from_range_result;
+
+	test->fa->falloc_from_range(test->fa, (__paddr_t *)test->start, 1, 0,
+					test->start + 4097,
+					test->start + 8192);
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Broken indentation

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+						&(test_attributes->start),
+						test_attributes->fa_meta_size,
+						&(test_attributes->pages));
+	UK_ASSERT(test_attributes->fa);
+
+	return test_attributes;
+
+}
+
+static void free_memory(struct test_ukfallocbuddy *test_attr)
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: It makes sense to match the function's name with the allocation function.
```suggestion
static void free_allocator(struct test_ukfallocbuddy *test_attr)
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 28]
```diff
@@ -0,0 +1,217 @@
...
+	struct uk_falloc *fa;
+
+	UK_ASSERT(fa = malloc(uk_fallocbuddy_size()));
+	UK_TEST_EXPECT_SNUM_EQ(uk_fallocbuddy_init(fa), 0);
+
+	free(fa);
+}
+
+/**
+ * Test if the addmem functionality works
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests if the addmem functionality works
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+	addmem_result = fa->addmem(fa, fa_meta, start, pages, dm_off);
+	UK_TEST_EXPECT_SNUM_EQ(addmem_result, 0);
+
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
+}
+
+/**
+ * Test variables have correct values
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests variables have correct values
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 30]
```diff
@@ -0,0 +1,217 @@
...
+	struct test_ukfallocbuddy *test = initialize_parameters(10000);
+
+	UK_TEST_EXPECT_SNUM_EQ(test->fa->free_memory, 8192);
+	UK_TEST_EXPECT_SNUM_EQ(test->fa->total_memory, 8192);
+
+	free_memory(test);
+}
+
+/**
+ * Test a simple allocation
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests a simple allocation
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+
+	falloc_result = test->fa->falloc(test->fa, (__paddr_t *)test->start,
+						1, 0);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_result, 0);
+
+	free_memory(test);
+}
+
+/**
+ * Test if alloc_from_range functionality works
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests if alloc_from_range functionality works
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 32]
```diff
@@ -0,0 +1,217 @@
...
+	falloc_from_range_result = test->fa->falloc_from_range(test->fa,
+					(__paddr_t *)test->start, 2, 0, min,
+					max);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, 0);
+
+	free_memory(test);
+}
+
+/**
+ * Test allocation with too many pages
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests allocation with too many pages
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 33]
```diff
@@ -0,0 +1,217 @@
...
+
+	falloc_result = test->fa->falloc(test->fa, (__paddr_t *)test->start,
+						600, 0);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_result, -EFAULT);
+
+	free_memory(test);
+}
+
+/**
+ * Test allocation over an already allocated zone
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests allocation over an already allocated zone
```

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c [hunk 34]
```diff
@@ -0,0 +1,217 @@
...
+					(__paddr_t *)test->start, 2, 0,
+					test->start, end);
+
+	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, -ENOMEM);
+
+	free_memory(test);
+}
+
+/**
+ * Test allocation of two frames between an already allocated one
```
Comment (marcrittinghaus) [inline lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]:
```suggestion
 * Tests allocation of two frames between an already allocated one
```

**Review by marcrittinghaus — CHANGES_REQUESTED** [review]
Hi @maniatro111,

please see the comments. Thanks!

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 34 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Changed.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 33 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Changed.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 32 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Changed.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 31 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Changed.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 30 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Changed.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 21 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Changed.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 23 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Thanks for noticing! Changed!

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 27 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Yes. It makes sense. Went with your suggestion.

File: lib/ukfallocbuddy/tests/test_ukfallocbuddy.c (hunk 20 above)
Comment (flpostolache) [reply lib/ukfallocbuddy/tests/test_ukfallocbuddy.c]: Went with your suggestion. Looks better now. Thank you!

**Review by Mihnea0Firoiu — APPROVED** [review]
Looks good to me.
Reviewed-by: Mihnea Firoiu <mihneafiroiu0@gmail.com>

**Review by razvanvirtan — COMMENTED** [review]
Hi, @flpostolache!
The PR looks good to me.

I am just thinking if we should make [this](https://github.com/unikraft/unikraft/pull/450/files#diff-f15384f071fd846efc5c9612e024214d2f8e475133b5ad951264f5bc35587de7R13) available to all tests, by moving it to `uktest` and eliminating the `fab` from the name, as it doesn't seem to have anything specific to the frame allocator.
WDYT?
